### ISSUE TITLE
Do not keep interfaces separated from implementations

### ISSUE THREAD

**Issue by samdark** [yiisoft/di#34]
- https://github.com/yiisoft/di/tree/master/src/contracts
- https://github.com/yiisoft/di/tree/master/src/support

Should be kept together.

**Comment by prowwid** [comment]
Hello @samdark ,

Why so? 

**contracts** directory contains only interfaces and in other words - abstraction. 
**support**, on the other hand,  contains only classes and in other words - default realization of abstraction defined in **contracts** . 

The point of having abstraction and realization separate is because you can have one abstraction with 1000 realizations and all these 1000 realizations would be aware of abstraction but the abstraction, at the same time, must not be aware of its realizations.

In other words, **contracts** contains a list of public interfaces everyone can use to develop realizations they need to use with container. But **support** contains only a list of default realizations that should be suited most of the cases.

**Comment by samdark** [comment]
Yes, that's true but there's no sense in having it like that since there aren't many and it's a single library anyway. End user may use interfaces regardless of where they are.

**Comment by SamMousa** [comment]
I am in favor of separating them directory wise and namespace wise, that's easier for coding in my opinion.
I am also in favor of having them in 1 packagist package and 1 repo, since it allows for easy development.

In my opinion putting them in the same directory doesn't really help anyone and basically forces us to append the `Interface` to the files to identify that they are interfaces not classes...

**Comment by samdark** [comment]
1 packagist package and 1 repo is no question. Closing since there seems to be general consensus about it and I don't mind too much.
